metric_tank/usage/usage.go

```diff
@@ -0,0 +1,107 @@
+package usage
+
+import (
+	"fmt"
+	"github.com/benbjohnson/clock"
+	"github.com/raintank/raintank-metric/metric_tank/defcache"
+	"github.com/raintank/raintank-metric/metric_tank/struc"
+	"github.com/raintank/raintank-metric/schema"
+	"sync"
+	"time"
+)
+
+var Clock clock.Clock
+var metrics struc.Metrics
+var defCache *defcache.DefCache
+
+type orgstat struct {
+	keys   map[string]struct{} // track unique keys seen
+	points uint32              // track amount of points seen
+}
+
+// tracks for every org
+type Usage struct {
+	sync.Mutex
+	period time.Duration
+	now    []orgstat
+	prev   []orgstat
+}
+
+func New(t time.Duration, m struc.Metrics, d *defcache.DefCache, cl clock.Clock) *Usage {
+	metrics = m
+	defCache = d
+	Clock = cl
+	return &Usage{
+		period: t,
+	}
+}
+
+func (u *Usage) Add(org int, key string) {
+	// org -1 is pos 0, etc.
+	// org 5 -> pos 6 -> we need len 7 or more -> len should be >= pos+1
+	pos := org + 1
+
+	u.Lock()
+	for len(u.now) < pos+1 {
```

Comment:
this logic looks flawed to me.
You are increasing the size of the u.now slice by 1 everytime that an orgId is greater then the number of previously seen orgs.

So lets say that the first org to report usage is orgId 1000.  u.now will have 1 orgstat added to it, making its length=1.  You then call `u.now[1001].points += 1` which will cause a panic due to index out of range.

as already mentioned above, just use a `map[int]orgstat`